backend/src/main/java/com/bikes/backend/service/MongoUserDetailsService.java

```diff
@@ -0,0 +1,38 @@
+package com.bikes.backend.service;
+
+import com.bikes.backend.model.MongoUser;
+import com.bikes.backend.repository.MongoUserRepository;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.User;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class MongoUserDetailsService implements UserDetailsService {
+	private final MongoUserRepository repository;
+
+	public MongoUserDetailsService(MongoUserRepository repository) {
+		this.repository = repository;
+	}
+
+	@Override
+	public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+		MongoUser mongoUser = repository.findByUsername(username)
+				.orElseThrow(() -> new UsernameNotFoundException("User not found"));
+
+		return new User(mongoUser.username(), mongoUser.password(),
+				List.of(new SimpleGrantedAuthority("ROLE_" + mongoUser.role())));
+	}
+
+	public Optional<MongoUser> loadMongoUserByUsername(String username) throws UsernameNotFoundException {
+		return Optional.ofNullable(repository.findByUsername(username))
+				.orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User not found"));
```

Comment:
Instead of unpacking the returned Optional and then re-packing it into a new Optional, you can directly return the Optional from your repository.